storage/watchable_store.go

```diff
@@ -241,57 +243,133 @@ func (s *watchableStore) syncWatchingsLoop() {
 	}
 }
 
-// syncWatchings syncs the watchings in the unsyncd map.
-func (s *watchableStore) syncWatchings() {
+// rangeUnsynced ranges on all unsynced watchings and returns all
+// revision bytes, key-value pairs and next revision.
+func (s *watchableStore) rangeUnsynced() ([][]byte, []storagepb.KeyValue, int64, map[string]map[*watching]struct{}, error) {
+	// get minimum and maximum revision to get all key-value pairs in unsynced
+	minRev, maxRev := int64(math.MaxInt64), int64(math.MinInt64)
+	keyToWatchings := make(map[string]map[*watching]struct{})
 	_, curRev, _ := s.store.Range(nil, nil, 0, 0)
 	for w := range s.unsynced {
-		var end []byte
-		if w.prefix {
-			end = make([]byte, len(w.key))
-			copy(end, w.key)
-			end[len(w.key)-1]++
-		}
-		limit := cap(w.ch) - len(w.ch)
-		// the channel is full, try it in the next round
-		if limit == 0 {
+		if w.cur > 0 && w.cur <= s.store.compactMainRev {
+			// TODO: handle this ErrCompacted error
+			log.Printf("storage: %v", ErrCompacted)
+			delete(s.unsynced, w)
 			continue
 		}
-		revbs, kvs, nextRev, err := s.store.RangeHistory(w.key, end, int64(limit), w.cur)
-		if err != nil {
-			// TODO: send error event to watching
+		if w.cur > s.store.currentRev.main {
+			// TODO: handle this ErrFutureRev error
+			log.Printf("storage: %v", ErrFutureRev)
 			delete(s.unsynced, w)
 			continue
 		}
 
-		// push events to the channel
-		for i, kv := range kvs {
-			var evt storagepb.Event_EventType
-			switch {
-			case isTombstone(revbs[i]):
-				evt = storagepb.DELETE
-			default:
-				evt = storagepb.PUT
+		k := string(w.key)
+
+		if w.cur > curRev {
+			if err := unsafeAddWatching(&s.synced, k, w); err != nil {
+				log.Panicf("error unsafeAddWatching (%v) for key %s", err, k)
 			}
+			delete(s.unsynced, w)
+			continue
+		}
+
+		if minRev >= w.cur {
+			minRev = w.cur
+		}
+		if maxRev <= w.cur {
+			maxRev = w.cur
+		}
+
+		if v, ok := keyToWatchings[k]; !ok {
+			keyToWatchings[k] = make(map[*watching]struct{})
+			keyToWatchings[k][w] = struct{}{}
+		} else {
+			v[w] = struct{}{}
+		}
+	}
+
+	// to prevent returning nothing
+	if minRev == maxRev {
+		maxRev++
+	}
+
+	min, max := newRevBytes(), newRevBytes()
+	revToBytes(revision{main: minRev}, min)
+	revToBytes(revision{main: maxRev, sub: maxRev}, max)
+
+	s.store.mu.Lock()
+	defer s.store.mu.Unlock()
+
+	// UnsafeRange returns keys and values. And in boltdb, keys are revisions.
+	tx := s.store.b.BatchTx()
+	tx.Lock()
+	ks, vs := tx.UnsafeRange(keyBucketName, min, max, 0)
+	tx.Unlock()
+
+	revbs := [][]byte{}
+	kvs := []storagepb.KeyValue{}
+	for i, v := range vs {
+		var kv storagepb.KeyValue
+		if err := kv.Unmarshal(v); err != nil {
+			return nil, nil, 0, nil, fmt.Errorf("storage: cannot unmarshal event: %v", err)
+		}
+		if _, ok := keyToWatchings[string(kv.Key)]; !ok {
+			continue
+		}
+		revbs = append(revbs, ks[i])
```

Comment:
i think we should move line 340 - 369 right here. Then we do not need to append these key-value pairs to a slice. And the code would be easier to read.